akka-stream/src/main/scala/akka/stream/impl/fusing/GraphInterpreter.scala

```diff
@@ -0,0 +1,434 @@
+/**
+ * Copyright (C) 2015 Typesafe Inc. <http://www.typesafe.com>
+ */
+package akka.stream.impl.fusing
+
+import akka.stream.stage.{ OutHandler, InHandler, GraphStage, GraphStageLogic }
+import akka.stream.{ Shape, Inlet, Outlet }
+
+/**
+ * INTERNAL API
+ *
+ * (See the class for the documentation of the internals)
+ */
+private[stream] object GraphInterpreter {
+  /**
+   * Compile time constant, enable it for debug logging to the console.
+   */
+  final val Debug = false
+
+  /**
+   * Marker object that indicates that a port holds no element since it was already grabbed. The port is still pullable,
+   * but there is no more element to grab.
+   */
+  case object Empty
+
+  sealed trait ConnectionState
+  sealed trait CompletedState extends ConnectionState
+  case object Pushable extends ConnectionState
+  case object Completed extends CompletedState
+  final case class PushCompleted(element: Any) extends ConnectionState
+  case object Cancelled extends CompletedState
+  final case class Failed(ex: Throwable) extends CompletedState
+
+  val NoEvent = -1
+  val Boundary = -1
+
+  abstract class UpstreamBoundaryStageLogic[T] extends GraphStageLogic {
+    def out: Outlet[T]
+  }
+
+  abstract class DownstreamBoundaryStageLogic[T] extends GraphStageLogic {
+    def in: Inlet[T]
+  }
+
+  /**
+   * INTERNAL API
+   *
+   * A GraphAssembly represents a small stream processing graph to be executed by the interpreter. Instances of this
+   * class **must not** be mutated after construction.
+   *
+   * The arrays [[ins]] and [[outs]] correspond to the notion of a *connection* in the [[GraphInterpreter]]. Each slot
```

Comment:
@drewhk what are these actually linking to? of course they will not work if they are not linking to something public. These should probably be changed to simple backticks.